smart-contract-feature-development.md

```diff
@@ -0,0 +1,94 @@
+# Purpose
+
+The purpose of this document is to provide guidance on how develop smart contracts without the need for feature branches.
+
+# Summary
+
+Feature branches result in a lot of effort spent keeping the feature branch up to date. We want to have modular code
+that enables us to easily compose features together such that we can get code mainlined early and toggle it off when
+running in production.
+
+# Problem Statement + Context
+
+We want to move away from a world with long lived feature branches.
+To do so, we need to have some flexibility when it comes to composition of features.
+This is easy in off-chain code as the cost is minimal for toggling off certain features but still having them in the source code.
+This is not the case for Solidity. This document exists to outline how to think about composition of features for Solidity code.
+When the feature is ready to be shipped, the feature specific contract should be merged into the base contract.
+
+# Alternatives Considered
+
+## Copy and Paste Implementation
+
+This was the solution for the fault proofs project, which worked well until the custom gas token feature was introduced. Now the contracts
+are in a state such that the custom gas token feature needs to be ported to the `OptimismPortal2` before the `OptimismPortal` and the
+`OptimismPortal2` can be merged. This may be a good solution for security critical code as we do not rely on the correctness
+of solidity inheritance and all of the logic lives in a single contract, making it easier to audit.
+
+# Proposed Solution
+
+Below includes general guidelines for how to develop solidity feature contracts that are meant to be merged into mainline at
+some point in the future.
+
+### Composition through Inheritance
+
+Solidity allows contacts to [inherit](https://docs.soliditylang.org/en/v0.8.17/contracts.html#inheritance) other contracts.
```

Comment:
Do we expect that this will make compile times worse/cause problems with the shift to `--via-ir`?